### ISSUE TITLE
Feature Request: High Level Architecture

### ISSUE THREAD

**Issue by RichCrook** [cypress-io/cypress-documentation#872]
I gave a well received talk about Cypress and have worked with it a bit, read the docs but are there any good sources for the architecture of Cypress? I know it's an Electron application but how specifically replace Selenium? It does not use Puppet I think...

If I'm asking folks to move away from Selenium, how is it actually implemented? I'd love to be able to give specifics with diagrams too!

Gleb Bahmutov not really, I think, but open an issue in cypress-documentation repo, that would be useful, even for high level architecture explanation

=========

Thanks,
Rich Crook

**Comment by richdouglasevans** [comment]
Spot on Rich!

I've been wanting this very same thing and have been messily doodling the emerging architecture as I've been reading the Cypress source. I don't have anything close to a polished set of diagrams yet but when I do I'll tag this ticket in the PR.

**Comment by jennifer-shehane** [comment]
Yeah there was this issue for this: https://github.com/cypress-io/cypress-documentation/issues/520

But, will close that one in favor of this.

**Comment by jennifer-shehane** [comment]
I made a really sad wireframe diagram that I never finalized back then 😢 

![screen shot 2018-08-29 at 11 33 40 am](https://user-images.githubusercontent.com/1271364/44798432-795af300-ab7f-11e8-916b-aa77ba6d602b.png)

**Comment by richdouglasevans** [comment]
I'll pick this one up then 🚀

> It'll be a couple of weeks at least since we've got a lot of work on at the mo'... so no PRs from me for a bit.
>
> Also, if Cypress is about to drop a release with a whole new-and-improved architecture lemme know before I scribble my way through this box of wax crayons 😆

![](https://media.giphy.com/media/zaezT79s3Ng7C/giphy.gif)

**Comment by jennifer-shehane** [comment]
No, no large architecture changes for a while.

**Comment by tlqiao** [comment]
@richdouglasevans , Is there any new update for high level architecture of cypress?

**Comment by richdouglasevans** [comment]
Yep, it's a work in progress... been doing it off and on as time permits. (Yeah, been busy with work.) I plan to complete it and post it up for review over the Xmas holidays when I'll have some free-er time.

So, this year for sure 😄

**Comment by TechGuyTy** [comment]
Curious if we ever got around to getting a finalized version published somewhere. I'm in the midst of googling away to see if one exists. I'm doing a presentation at my company to try and convince the team that this is a worthwhile investment (I know it is, just need to convince some other folks) and having a diagram of the architecture would be nice to have.

**Comment by RichCrook** [comment]
I’ve neither seen nor heard an update around this as of yet but I hope something soon...

**Comment by Light07** [comment]
Do we have any update available for now?

**Comment by ayandebbarman** [comment]
Following this as I am looking for same. Lot of Selenium flakiness comes from the way it interacts with browser so i am making a case to others to use Cypress but couldn't point to documentation which explains how Cypress will not have the same issue

**Comment by shankarnarayanb** [comment]
I found the diagram from Dominic Elm quite useful :-)

![Picture1](https://user-images.githubusercontent.com/17179877/74605728-8e547a80-50c2-11ea-8549-b804ef9b4996.png)

**Comment by Light07** [comment]
@shankarnarayanb ， Thanks so much for this,  it's useful!

**Comment by Pranav-Mahale** [comment]
> I found the diagram from Dominic Elm quite useful :-)
> 
> ![Picture1](https://user-images.githubusercontent.com/17179877/74605728-8e547a80-50c2-11ea-8549-b804ef9b4996.png)

@shankarnarayanb, Instead of referring any other doc this 'Cypress Architectural Diagram' looks very neat and clear.
I'm new to the Cypress so unable to get in-depth. It would be very helpful if you can brief about this 'Cypress Architectural Diagram'

Thanks,
Pranav

**Comment by lucas-nguyen-17** [comment]
As my understanding, the workflow will be like:
1. There is a nodejs server, which launches a browser with a proxy.
2. The browser shows up with 2 iframes: 1 for cypress, 1 for application. Iframe of application may be injected a script to change the domain to localhost, then 2 iframes can talk to each other. As a result, cypress can access all the objects like DOM, Window, LocalStorage... of application.
3. Cypress and Nodejs need communicate to each other to ask for privileges of OS, such as take screenshots or record video...
4. Browser starts with proxy, it means that Nodejs control the network layer. Therefore, Nodejs can leverage this point to capture request then change it back and forth.

**Comment by GabRog** [comment]
Hello :)
I am using now Protractor, that in docs has clear architecture schema provided. 
I tried writing some tests in Cypress, I wanted to get in-depth and make architecture comparison between these two tools. 
However, there are still no details of Cypress architecture provided. 
This issue is opened for nearly three years, I cannot believe that you have no schema you can provide at once and keep updated. 
Could you prioritize this ticket and share such information with Cypress users? As you could notice above, there are some people interested not only in "how to write it to just run", but also "how it really works and why is it better from one more perspective than other tools". 
Please, give us one more argument for using your tool.

**Comment by avietlobo** [comment]
> Hello :)
> I am using now Protractor, that in docs has clear architecture schema provided.
> I tried writing some tests in Cypress, I wanted to get in-depth and make architecture comparison between these two tools.
> However, there are still no details of Cypress architecture provided.
> This issue is opened for nearly three years, I cannot believe that you have no schema you can provide at once and keep updated.
> Could you prioritize this ticket and share such information with Cypress users? As you could notice above, there are some people interested not only in "how to write it to just run", but also "how it really works and why is it better from one more perspective than other tools".
> Please, give us one more argument for using your tool.

Hi :)

Please refer below blog written by Jonas Verhoelen for further clarity on the architecture

https://blog.codecentric.de/en/2020/10/cypress-ui-end2end-testing/
